### PR TITLE
refactor (model): move from github.com/pkg/errors to 'errors' and 'fmt'

### PR DESCRIPTION
**Pull request by mmorel-35**
The package https://github.com/pkg/errors is not maintained anymore.
This PR replaces https://github.com/pkg/errors with official errors and fmt golang packages in model package

Signed-off-by: Matthieu MOREL [mmorel-35@users.noreply.github.com](mailto:mmorel-35@users.noreply.github.com)

### CODE REVIEW

File: model/rulefmt/rulefmt_test.go [hunk 1]
```diff
@@ -182,8 +182,12 @@ groups:
 `
 	_, errs := Parse([]byte(group))
 	require.Len(t, errs, 2, "Expected two errors")
-	err0 := errs[0].(*Error).Err.node
-	err1 := errs[1].(*Error).Err.node
+	var err00 *Error
+	errors.As(errs[0], &err00)
+	err0 := err00.Err.node
+	var err01 *Error
+	errors.As(errs[1], &err01)
```
Comment (roidelapluie) [inline model/rulefmt/rulefmt_test.go:189]: same

File: model/rulefmt/rulefmt.go [hunk 2]
```diff
@@ -40,12 +41,15 @@ type Error struct {
 
 // Error prints the error message in a formatted string.
 func (err *Error) Error() string {
-	if err.Err.nodeAlt != nil {
-		return errors.Wrapf(err.Err.err, "%d:%d: %d:%d: group %q, rule %d, %q", err.Err.node.Line, err.Err.node.Column, err.Err.nodeAlt.Line, err.Err.nodeAlt.Column, err.Group, err.Rule, err.RuleName).Error()
-	} else if err.Err.node != nil {
-		return errors.Wrapf(err.Err.err, "%d:%d: group %q, rule %d, %q", err.Err.node.Line, err.Err.node.Column, err.Group, err.Rule, err.RuleName).Error()
+	if err.Err.err != nil {
+		if err.Err.nodeAlt != nil {
+			return fmt.Errorf("%d:%d: %d:%d: group %q, rule %d, %q: %w", err.Err.node.Line, err.Err.node.Column, err.Err.nodeAlt.Line, err.Err.nodeAlt.Column, err.Group, err.Rule, err.RuleName, err.Err.err).Error()
```
Comment (roidelapluie) [inline model/rulefmt/rulefmt.go:46]: Can we do with Sprintf directly?

File: model/textparse/promlex.l.go [hunk 3]
```diff
@@ -16,7 +16,7 @@
 package textparse
 
 import (
-	"github.com/pkg/errors"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (mmorel-35) [reply model/textparse/promlex.l.go:19]: It is not generated, directly modified

File: model/textparse/promlex.l [hunk 4]
```diff
@@ -27,6 +27,9 @@ const (
     sLValue
     sValue
     sTimestamp
+    sExemplar
+    sEValue
+    sETimestamp
```
Comment (roidelapluie) [inline model/textparse/promlex.l:32]: Examplars are not supported by the prometheus text format

Comment (mmorel-35) [reply model/textparse/promlex.l:32]: But they are needed for the model/textparse/openmetricslex.l.go and the consts cannot be declared twice in the same package . How should it be handled ?

Comment (roidelapluie) [reply model/textparse/promlex.l:32]: ooh I see

Comment (roidelapluie) [reply model/textparse/promlex.l:32]: it seems we have diverted from the auto generated files

**Comment by mmorel-35** [comment]
Hi @roidelapluie , do you see anything left to do here ?

**Comment by mmorel-35** [comment]
Hi @bboreham ,

> IMHO this is not a refactor; the behaviour of `fmt.Errorf` is substantially different to `errors.Errorf`. (To be clear I am complaining about the description, not the effect of this PR.)

How would you name it ? The intention is to keep the same behavior while using the standard libraries instead of deprecated and archived one.

> Also you changed `WrappedError` to tolerate wrapping `nil`; would be nice to see an explanation why that is necessary.

If https://pkg.go.dev/github.com/pkg/errors#Wrap receives a nil error it returns nil so if errors.Wrap(err, "..").Error() is called it will trigger a nil pointer. Would it be possible and even acceptable to trigger the same error while moving to errors and fmt ?

File: model/rulefmt/rulefmt.go [hunk 5]
```diff
@@ -40,12 +41,16 @@ type Error struct {
...
 // Error prints the error message in a formatted string.
 func (err *Error) Error() string {
+	if err.Err.err == nil {
+		return ""
+	}
 	if err.Err.nodeAlt != nil {
-		return errors.Wrapf(err.Err.err, "%d:%d: %d:%d: group %q, rule %d, %q", err.Err.node.Line, err.Err.node.Column, err.Err.nodeAlt.Line, err.Err.nodeAlt.Column, err.Group, err.Rule, err.RuleName).Error()
-	} else if err.Err.node != nil {
-		return errors.Wrapf(err.Err.err, "%d:%d: group %q, rule %d, %q", err.Err.node.Line, err.Err.node.Column, err.Group, err.Rule, err.RuleName).Error()
+		return fmt.Sprintf("%d:%d: %d:%d: group %q, rule %d, %q: %v", err.Err.node.Line, err.Err.node.Column, err.Err.nodeAlt.Line, err.Err.nodeAlt.Column, err.Group, err.Rule, err.RuleName, err.Err.err)
```
Comment (roidelapluie) [inline model/rulefmt/rulefmt.go:48]:
```suggestion
		return fmt.Sprintf("%d:%d: %d:%d: group %q, rule %d, %q: %w", err.Err.node.Line, err.Err.node.Column, err.Err.nodeAlt.Line, err.Err.nodeAlt.Column, err.Group, err.Rule, err.RuleName, err.Err.err)
```

File: model/rulefmt/rulefmt.go [hunk 6]
```diff
@@ -40,12 +41,16 @@ type Error struct {
...
+	}
+	if err.Err.node != nil {
+		return fmt.Sprintf("%d:%d: group %q, rule %d, %q: %v", err.Err.node.Line, err.Err.node.Column, err.Group, err.Rule, err.RuleName, err.Err.err)
```
Comment (roidelapluie) [inline model/rulefmt/rulefmt.go:51]:
```suggestion
		return fmt.Sprintf("%d:%d: group %q, rule %d, %q: %w", err.Err.node.Line, err.Err.node.Column, err.Group, err.Rule, err.RuleName, err.Err.err)
```

File: model/rulefmt/rulefmt.go [hunk 7]
```diff
@@ -40,12 +41,16 @@ type Error struct {
...
 	}
-	return errors.Wrapf(err.Err.err, "group %q, rule %d, %q", err.Group, err.Rule, err.RuleName).Error()
+	return fmt.Sprintf("group %q, rule %d, %q: %v", err.Group, err.Rule, err.RuleName, err.Err.err)
```
Comment (roidelapluie) [inline model/rulefmt/rulefmt.go:53]:
```suggestion
	return fmt.Sprintf("group %q, rule %d, %q: %w", err.Group, err.Rule, err.RuleName, err.Err.err)
```

File: model/rulefmt/rulefmt.go [hunk 8]
```diff
@@ -58,10 +63,14 @@ type WrappedError struct {
...
 // Error prints the error message in a formatted string.
 func (we *WrappedError) Error() string {
+	if we.err == nil {
+		return ""
+	}
 	if we.nodeAlt != nil {
-		return errors.Wrapf(we.err, "%d:%d: %d:%d", we.node.Line, we.node.Column, we.nodeAlt.Line, we.nodeAlt.Column).Error()
-	} else if we.node != nil {
-		return errors.Wrapf(we.err, "%d:%d", we.node.Line, we.node.Column).Error()
+		return fmt.Sprintf("%d:%d: %d:%d: %v", we.node.Line, we.node.Column, we.nodeAlt.Line, we.nodeAlt.Column, we.err)
```
Comment (roidelapluie) [inline model/rulefmt/rulefmt.go:70]:
```suggestion
		return fmt.Sprintf("%d:%d: %d:%d: %w", we.node.Line, we.node.Column, we.nodeAlt.Line, we.nodeAlt.Column, we.err)
```

File: model/textparse/openmetricslex.l.go [hunk 9]
```diff
@@ -747,16 +735,58 @@ yyrule25: // {S}[^ \n]+
...
 		return tLinebreak
 		goto yystate0
 	}
 	panic("unreachable")
 
-	goto yyabort // silence unused label error
-
 yyabort: // no lexem recognized
+	//
+	// silence unused label errors for build and satisfy go vet reachability analysis
```
Comment (roidelapluie) [inline model/textparse/openmetricslex.l.go:747]: ?? What created this changes in generated file?

File: model/textparse/promlex.l.go [hunk 10]
```diff
@@ -521,16 +512,50 @@ yyrule18: // {D}+
...
 		return tLinebreak
 		goto yystate0
 	}
 	panic("unreachable")
 
-	goto yyabort // silence unused label error
-
 yyabort: // no lexem recognized
+	//
+	// silence unused label errors for build and satisfy go vet reachability analysis
```
Comment (roidelapluie) [inline model/textparse/promlex.l.go:524]: is it generated by golex?

File: model/rulefmt/rulefmt.go (hunk 8 above)
Comment (mmorel-35) [reply model/rulefmt/rulefmt.go:70]: Are you sure ? I have this warning in vscode : 
> fmt.Sprintf does not support error-wrapping directive %w

File: model/textparse/openmetricslex.l.go (hunk 9 above)
Comment (mmorel-35) [reply model/textparse/openmetricslex.l.go:747]: I executed the recommended command from [here](https://github.com/prometheus/prometheus/tree/main/model/textparse#making-changes-to-textparse-lexers) and that is the result I got. Don't you ?

File: model/textparse/promlex.l.go (hunk 10 above)
Comment (mmorel-35) [reply model/textparse/promlex.l.go:524]: Exactly !

**Comment by bboreham** [comment]
> How would you name it ? 

"Move from X to Y" is fine.

> The intention is to keep the same behavior while using the standard libraries instead of deprecated and archived one.

OK, but you have to pick. They don't have the same behaviour.

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
On 17 Jun 05:12, Matthieu MOREL wrote:
> @mmorel-35 commented on this pull request.
> 
> 
> 
> >  	if we.nodeAlt != nil {
> -		return errors.Wrapf(we.err, "%d:%d: %d:%d", we.node.Line, we.node.Column, we.nodeAlt.Line, we.nodeAlt.Column).Error()
> -	} else if we.node != nil {
> -		return errors.Wrapf(we.err, "%d:%d", we.node.Line, we.node.Column).Error()
> +		return fmt.Sprintf("%d:%d: %d:%d: %v", we.node.Line, we.node.Column, we.nodeAlt.Line, we.nodeAlt.Column, we.err)
> 
> Are you sure ? I have this warning in vscode : 
> > fmt.Sprintf does not support error-wrapping directive %w


oooh indeed. What about %s?

> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/prometheus/prometheus/pull/10747#discussion_r900058987
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

-- 
Julien Pivotto
O11y - https://o11y.eu/

**Pull request merged**